### PR TITLE
[WIP] Build: remove npm-shrinkwrap.json and try yarn

### PR DESCRIPTION
**Pull request by gwwar**
An in progress branch to see if we can use yarn and ditch npm-shrinkwrap.json

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -32,9 +33,9 @@ ENV     NODE_PATH /calypso/server:/calypso/client
 
 # Install base npm packages to take advantage of the docker cache
 COPY    ./package.json /calypso/package.json
-COPY    ./npm-shrinkwrap.json /calypso/npm-shrinkwrap.json
-# Sometimes "npm install" fails the first time when the cache is empty, so we retry once if it failed
-RUN     npm install --production || npm install --production
+COPY    ./yarn.lock /calypso/yarn.lock
+
+RUN     curl -o- -L https://yarnpkg.com/install.sh | bash && export PATH="$HOME/.yarn/bin:$PATH" && yarn install
```
Comment (gwwar) [inline Dockerfile]: Probably better to wait until @dmsnell's docker changes are in before debugging this too far

**Comment by gwwar** [comment]
Odd, one failing test with yarn vs npm.

screen shot 2016-10-11 at 3 48 22 pm

File: Dockerfile (hunk 1 above)
Comment (DanReyLop) [inline Dockerfile]: You can just install it using `npm install yarn --global` . Somehow I find the `curl | bash` installation method too magic :)

Comment (umurkontaci) [inline Dockerfile]: How about `npm install yarn=0.15.1 -g --prefix ./node_utils` and then run with `./node_utils/bin/yarn`?

This way we can control the version we are using and not clutter global.

Comment (dmsnell) [inline Dockerfile]: whoa! where's that global flag coming from - stop messing up my system! :smile:

@gwwar in relation to the Dockerfile we may have more issues with the deployment process than code conflicts because it specifically sets a local npm mirror.

Comment (gwwar) [inline Dockerfile]: I mostly meant that we should wait since our base image is changing from Debian to node-slim with @dmsnell upcoming changes. If you look at https://yarnpkg.com/en/docs/install#linux they also provide the package via apt-get, but I was running into trouble configuring the apt-get repository to be able to download yarn.

Comment (gwwar) [inline Dockerfile]: Note to self, this should also be run with `yarn install --production` to skip over devDependencies. https://yarnpkg.com/en/docs/cli/install

Comment (dmsnell) [inline Dockerfile]: > yarn install --production to skip over devDependencies

probably not @gwwar because the Docker container builds the bundles from source. we may not need linting but we certainly need the loaders and babel etc…

Comment (gwwar) [inline Dockerfile]: If we use something in both prod/dev, that makes it a plain dependency 😄 If you look at our package.json, babel should be under `dependencies` vs `devDependencies`

**Comment by gwwar** [comment]
Hmm got the build green, but running `make test` locally for me causes node to run out of memory.
screen shot 2016-10-12 at 2 12 18 pm

File: Makefile [hunk 2]
```diff
@@ -80,7 +81,10 @@ welcome:
 	@printf "\033[36m               |___/|_|                  \n"
 	@printf "\033[m\n"
 
-install: node_modules
+yarn_available:
```
Comment (dmsnell) [inline Makefile]: this seems like an odd name for a make target. it's almost as if it should return some boolean indication, but it's really just forcing a dependency. maybe something like `demand_yarn` or `yarn_check` would fit better

**Comment by dmsnell** [comment]
@gwwar why do we need yarn installed globally? is it possible to use a local copy and `npm install yarn` before building? personally I've dealt with plenty of headaches in the past two months because of global version clashes and that has been so frustrating to resolve

File: circle.yml [hunk 3]
```diff
@@ -1,6 +1,11 @@
 machine:
   node:
     version: 6.7.0
+dependencies:
+  pre:
+    - curl -o- -L https://yarnpkg.com/install.sh | bash
```
Comment (dmsnell) [inline circle.yml]: versioning much? I think we could pull from https://github.com/yarnpkg/yarn/releases if we wanted to pin it and make updates their own PRs

**Comment by gwwar** [comment]
> @gwwar why do we need yarn installed globally?

I can give that a shot. Was doing some quick hacks yesterday ;D

File: circle.yml (hunk 3 above)
Comment (gwwar) [inline circle.yml]: Hmm interesting thought for fixing yarn versions. Should we maybe try to bump to latest on each node bump?

Comment (dmsnell) [inline circle.yml]: Seems good to me. Having it versioned makes for nice explicit changes in PRs

File: client/lib/purchases/test/assembler.js [hunk 4]
```diff
@@ -32,18 +32,17 @@ describe( 'assembler', () => {
...
+		const creditCard = payment.creditCard;
+		expect( creditCard.expiryDate ).to.equal( '11/16' );
+		expect( creditCard.expiryMoment.isSame( moment( '11/16', 'MM/YY' ) ) ).to.equal( true );
+		expect( creditCard.id ).to.equal( 1234 );
+		expect( creditCard.number ).to.equal( 7890 );
+		expect( creditCard.type ).to.equal( 'visa' );
+		expect( payment.type ).to.equal( 'credit_card' );
+		expect( payment.countryCode ).to.equal( 'US' );
+		expect( payment.countryName ).to.equal( 'United States' );
+		expect( payment.name ).to.equal( 'My VISA' );
```
Comment (gwwar) [inline client/lib/purchases/test/assembler.js]: added in https://github.com/Automattic/wp-calypso/pull/8698

**Comment by dmsnell** [comment]
> Edit: note that we'll see this annoying prompt npm WARN prefer global yarn@0.15.1 should be installed with -g since "preferGlobal": true, is set in their package.json

yuck. it's probably okay if we need to. I see the causal dependence on npm that could make this tricky and justify a global. globals can be just so nasty

**Comment by umurkontaci** [comment]
I said this above but got no response, what about:

`npm install yarn=0.15.1 -g --prefix ./node_utils` and then run with `./node_utils/bin/yarn`? AFAIK, `--global` puts an executable inside the `bin` folder for node. If we use `--prefix` along with `--global`, it will install everything inside the folder we specify, without leaking.

We can then `.gitignore` that folder and control the yarn version from Makefile.

**Comment by dmsnell** [comment]
sorry if I missed your comment @umurkontaci but that sounds like a great idea

now, let's pause and consider our other globals. how many (if any) do we depend on? could this same solution make all of them non-global?

**Comment by umurkontaci** [comment]
> how many (if any) do we depend on? could this same solution make all of them non-global?

Apart from shonkwrap, I don't think we have any.

```
$ npm ls -g --depth=0 2>/dev/null
/usr/local/lib
`-- npm@3.10.3
```

After `make` && `make run` on fresh docker image `node:6.7.0`

**Comment by gwwar** [comment]
Just FYI  It's poorly documented at the moment, but @blowery mentioned in the hangout that yarn stores a cache at `~/.yarn-cache` which is why it suggests the global flag.

@umurkontaci I think that approach looks good, but we'll want to check where it's placing the cache directory

**Comment by umurkontaci** [comment]
Good catch. Per https://github.com/yarnpkg/yarn/blob/master/bin/yarn.js#L11 it will use the home directory of the user regardless. 

Home directory is `$HOME` in Unix and `$USERPROFILE` in Win32. [*](http://docs.libuv.org/en/v1.x/misc.html#c.uv_os_homedir) If we really want, we can set `$HOME` to a different value while calling yarn. I don't see this being a huge issue as far as yarn is concerned, but it makes me very uncomfortable to change `$HOME`.

We can probably fix this in the upstream by adding an option to set the cache file name from an env value or an optional command.

**Comment by gwwar** [comment]
> We can probably fix this in the upstream by adding an option to set the cache file name from an env value or an optional command.

Thanks for 👀 @umurkontaci. That seems like a good step to take.

We can probably steal a few ideas from https://github.com/yarnpkg/yarn/issues/749 as well

**Comment by dmsnell** [comment]
@gwwar any reason this is on hold? are we waiting for more stabilization in yarn? I think we can :ship: it with `yarn` installed globally. it's still a net positive.

**Comment by gwwar** [comment]
I'm waiting for yarn to stabilize a bit more. There's open issues like https://github.com/yarnpkg/yarn/issues/761 that I'd not like to debug in prod
